### PR TITLE
Connectivity graph

### PR DESCRIPTION
**Pull request by Zshan0**
Functionality for accessing connectivity graph from IBM backend. Since accessing the actual hardware without gaining a subscription is impossible, I have created tests that fake the hardware responses.

### CODE REVIEW

File: packages/qiskit/quri_parts/qiskit/backend/connectivity.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import networkx as nx
+from qiskit.providers.ibmq import IBMQBackend
+
+
+def device_connectivity_graph(device: IBMQBackend) -> nx.Graph:
```
Comment (kwkbtr) [inline packages/qiskit/quri_parts/qiskit/backend/connectivity.py]: I wonder if we can use this function with a backend more general than `IBMQBackend`.
In `QiskitSamplingBackend`, `BackendV1` is used when accessing `backend.configuration()`:
https://github.com/QunaSys/quri-parts/blob/4e739e0fd7e21f4734e4cc742dec97ce194774b8/packages/qiskit/quri_parts/qiskit/backend/sampling.py#L122
For `BackendV2`, it seems that `num_qubits` and `coupling_map` can be retrieved as direct attributes:
https://qiskit.org/documentation/stubs/qiskit.providers.BackendV2.html#backendv2

File: packages/qiskit/tests/qiskit/backend/test_qiskit_connectivity.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+class MockBackend:
+    def __init__(self, configuration: MockConfiguration) -> None:
+        self.config = configuration
+
+    def configuration(self) -> MockConfiguration:
+        return self.config
+
+
+class TestHTDiagonalization:
```
Comment (kwkbtr) [inline packages/qiskit/tests/qiskit/backend/test_qiskit_connectivity.py]: This class name should be renamed.

File: packages/qiskit/quri_parts/qiskit/backend/connectivity.py (hunk 1 above)
Comment (Zshan0) [reply packages/qiskit/quri_parts/qiskit/backend/connectivity.py]: Would you like me to add support for both `BackendV1` and `BackendV2`? 

### `BackendV1` Fetching

The `BackendV1` class has a `configuration` method that returns a `BackendConfiguration` ([link](https://qiskit.org/documentation/_modules/qiskit/providers/models/backendconfiguration.html#BackendConfiguration)). This object contains `num_qubits` and `coupling_map` attributes (refer to source [code](https://qiskit.org/documentation/_modules/qiskit/providers/models/backendconfiguration.html#BackendConfiguration)). `coupling_map` is of the type `CouplingMap` ([documentation](https://qiskit.org/documentation/stubs/qiskit.transpiler.CouplingMap.html)) with a `get_edges` method that returns the list of edges in the map. The required connectivity graph that we use can be constructed from this.


### `BackendV2` Fetching

The `BackendV2` class has a much simpler interface. I can use the `coupling_map` and `num_qubits` attributes directly and construct the required graph.

By doing instance checking, I could handle both of these cases. For testing purposes, I will modify the classes I have constructed to match the backend functionality in qiskit.

Comment (kwkbtr) [reply packages/qiskit/quri_parts/qiskit/backend/connectivity.py]: Yes, supporting both of them is good!

File: packages/qiskit/tests/qiskit/backend/test_qiskit_connectivity.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+        config = MockConfiguration(coupling_map, num_qubits)
+        backend = MockBackendV1(config)
+
+        received_graph = device_connectivity_graph(backend)
+
+        edge_list = [(0, 1)]
+        lines = [f"{a} {b}" for (a, b) in edge_list]
+
+        constructed_graph = nx.parse_edgelist(lines, nodetype=int)
```
Comment (kwkbtr) [inline packages/qiskit/tests/qiskit/backend/test_qiskit_connectivity.py]: I think this code works fine, but simply creating a graph as `nx.Graph([(0, 1)])` is easier to read.

File: packages/qiskit/tests/qiskit/backend/test_qiskit_connectivity.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        num_qubits = 2
+
+        backend = MockBackendV2(coupling_map, num_qubits)
+
+        received_graph = device_connectivity_graph(backend)
+
+        edge_list = [(0, 1)]
+        lines = [f"{a} {b}" for (a, b) in edge_list]
+
+        constructed_graph = nx.parse_edgelist(lines, nodetype=int)
```
Comment (kwkbtr) [inline packages/qiskit/tests/qiskit/backend/test_qiskit_connectivity.py]: I think this code works fine, but simply creating a graph as `nx.Graph([(0, 1)])` is easier to read.

**Review by kwkbtr — APPROVED** [review]
I suggested tiny improvements but I think it's good to go!

**Comment by Zshan0** [comment]
I have made the necessary changes and the tests are passing, if appropriate please merge it

**Review by kwkbtr — APPROVED** [review]

**Pull request merged**